File: challenge/leetcode/cpp/addone.cpp
#include <vector>
#include <iostream>

std::vector<int> addOne(std::vector<int> arr) {
    int carry = 1;
    int sum = 0;
    for (int i = arr.size() - 1; i >= 0; i--) {
        sum = arr[i] + carry;
        carry = sum / 10;
        arr[i] = sum % 10;
        if (carry == 0) {
            return arr;
        }
    }
    std::vector<int> newArr = std::vector<int>(arr.size() + 1);
    newArr[0] = carry;
    for (int i = 0; i < arr.size(); i++) {
        newArr[i + 1] = arr[i];
    }
    return newArr;
}

int main() {
    std::vector<int> arr{9};
    std::vector<int> newArr = addOne(arr);
    for (int i : newArr) {
        std::cout << i << ", ";
    }
}

File: challenge/leetcode/java/AddOne.java
public class AddOne {
    public static int[] addOne(int[] arr) {
        int carry = 1;
        int sum = 0;
        for (int i = arr.length - 1; i >= 0; i--) {
            sum = arr[i] + carry;
            carry = sum / 10;
            arr[i] = sum % 10;
            if (carry == 0)
                return arr;
        }
        int[] newArr = new int[arr.length + 1];
        newArr[0] = carry;
        for (int i = 0; i < arr.length; i++) {
            newArr[i + 1] = arr[i];
        }
        return newArr;
    }

    public static int[] recurAddOne(int[] arr) {
        addOne(arr, arr.length - 1);
        if (arr[0] == 0) {
            int[] newArr = new int[arr.length + 1];
            newArr[0] = 1;
            for (int i = 0; i < arr.length; i++) {
                newArr[i + 1] = arr[i];
            }
            return newArr;
        }
        return arr;
    }

    private static void addOne(int[] arr, int index) {
        if (index < 0) {
            return;
        }
        if (arr[index] < 9) {
            arr[index] += 1;
            return;
        }
        arr[index] = 0;
        addOne(arr, index - 1);
    }

    public static void main(String[] args) {
        // int[] arr = new int[]{9};
        // int[] arr = new int[]{1, 2, 3};
        int[] arr = new int[]{9, 9, 9};
        // int[] newArr = addOne(arr);
        int[] newArr = recurAddOne(arr);
        for (int i = 0; i < newArr.length; i++) {
            System.out.print(newArr[i] + ", ");
        }
    }
}
